### PR TITLE
ANC Register 2019

### PR DESCRIPTION
**Pull request by solemabrothers**
https://app.asana.com/0/1133167201254915/1169532876406257

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        rd.setName(getName());
+        rd.setDescription(getDescription());
+        rd.addParameters(getParameters());
+        rd.addDataSetDefinition("ANC", Mapped.mapStraightThrough(dataSetDefinition()));
+        return rd;
+    }
+
+    @Override
+    public String getVersion() {
+        return "2.0.9.7";
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java:126]: @solema only 3 digits not 4, should be 3.0.0

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        dsd.addColumn("Mebendazole", sdd.definition("Mebendazole", sdd.getConcept("9d6abbc4-707a-4ec7-a32a-4090b1c3af87")), "onOrAfter=${startDate},onOrBefore=${endDate}", new MebendazoleDataConverter());
+        dsd.addColumn("Iron given", sdd.getIronGiven(), "onDate=${endDate}", new CalculationResultDataConverter());
+        dsd.addColumn("Folic acid given", sdd.getFolicAcidGiven(), "onDate=${endDate}", new CalculationResultDataConverter());
+        dsd.addColumn("Other treatments", sdd.definition("Other treatments", sdd.getConcept("2aa72406-436e-490d-8aa4-d5336148204f")), "onOrAfter=${startDate},onOrBefore=${endDate}", new ObsDataConverter());
+        dsd.addColumn("Referal In/Out", sdd.referredToOrFrom(), "onDate=${endDate}", new CalculationResultDataConverter());
+        dsd.addColumn("Risk Factor/Complications", sdd.definition("Risk Factor/Complications", sdd.getConcept("120186AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")), "onOrAfter=${startDate},onOrBefore=${endDate}", new ObsDataConverter());
+
+        return dsd;
+    }
+    private DataDefinition getEncounterDate() {
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java:219]: @solemabrothers This needs to be from a global definition

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java (hunk 1 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java]: done

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java (hunk 2 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java]: Done

Comment (ssmusoke) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java]: @solemabrothers You do not need a private method for this. 

Also what is the difference between ANCEncounterCalculation and other encounter calculations?

File: api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/Anc1TimingDataConverter.java [hunk 3]
```diff
@@ -28,10 +28,10 @@ public Object convert(Object obj) {
         }
         Obs obs = ((Obs) obj);
 
-        if (obs.getValueCoded() != null && obs.getValueCoded().equals(Dictionary.getConcept(Dictionary.YES_CIEL))) {
-            return "=UNICHAR(8730)";
+        if (obs.getValueCoded() != null && obs.getValueCoded().equals(Dictionary.getConcept("dcd695dc-30ab-102d-86b0-7a5022ba4115"))) {
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/Anc1TimingDataConverter.java:31]: @solemabrothers why the change from Dictionary.YES_CIEL? The old pre-2019 reports will now be broken.

Just add new checks rather than replacing the existing ones

File: api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/MNCDataConverter.java [hunk 4]
```diff
@@ -29,11 +29,11 @@ public Object convert(Object obj) {
         }
         Concept mncResults = ((Obs) obj).getValueCoded();
 
-        if(mncResults != null && mncResults.equals(Dictionary.getConcept(Dictionary.YES_CIEL))){
+        if(mncResults != null && mncResults.equals(Dictionary.getConcept("dcd695dc-30ab-102d-86b0-7a5022ba4115"))){
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/MNCDataConverter.java:32]: @solemabrothers  ANC1TimingDataConverter

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reporting/dataset/definition/SharedDataDefintion.java [hunk 5]
```diff
@@ -28,9 +28,7 @@
 import org.openmrs.module.reporting.data.person.definition.ObsForPersonDataDefinition;
 import org.openmrs.module.reporting.evaluation.parameter.Parameter;
 import org.openmrs.module.ugandaemrreports.definition.data.definition.CalculationDataDefinition;
-import org.openmrs.module.ugandaemrreports.reporting.calculation.anc.AgeLimitCalculation;
-import org.openmrs.module.ugandaemrreports.reporting.calculation.anc.PersonAddressCalculation;
-import org.openmrs.module.ugandaemrreports.reporting.calculation.anc.WhoCd4VLCalculation;
+import org.openmrs.module.ugandaemrreports.reporting.calculation.anc.*;
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reporting/dataset/definition/SharedDataDefintion.java:33]: @solemabrothers Remove wildcard characters

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+import org.openmrs.module.reporting.data.patient.definition.PatientIdentifierDataDefinition;
+import org.openmrs.module.reporting.data.person.definition.PersonAttributeDataDefinition;
+import org.openmrs.module.reporting.data.person.definition.PreferredNameDataDefinition;
+import org.openmrs.module.reporting.dataset.definition.DataSetDefinition;
+import org.openmrs.module.reporting.dataset.definition.PatientDataSetDefinition;
+import org.openmrs.module.reporting.evaluation.parameter.Mapped;
+import org.openmrs.module.reporting.evaluation.parameter.Parameter;
+import org.openmrs.module.reporting.report.ReportDesign;
+import org.openmrs.module.reporting.report.definition.ReportDefinition;
+import org.openmrs.module.ugandaemrreports.data.converter.*;
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java:21]: @solemabrothers remove all wildcard imports

File: api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/MNCDataConverter.java (hunk 4 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/MNCDataConverter.java]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reporting/dataset/definition/SharedDataDefintion.java (hunk 5 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reporting/dataset/definition/SharedDataDefintion.java:33]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java (hunk 6 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/Anc1TimingDataConverter.java (hunk 3 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/data/converter/Anc1TimingDataConverter.java]: Done

**Comment by ssmusoke** [comment]
@solemabrothers the report does not run error message here https://gist.github.com/ssmusoke/c64aca845c89b085b66cacc4415a8e63

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Override
+    public String getUuid() {
+        return "95b4a987-2a8d-41f2-94b2-9c5bbeef9ca9";
+    }
+
+    @Override
+    public String getName() {
+        return "Intergrated Antenatal Register";
```
Comment (ssmusoke) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java:108]: @solemabrothers add 2019 (just like other registers)

**Comment by ssmusoke** [comment]
@solemabrothers Error 

`java.lang.NullPointerException
	at org.openmrs.module.ugandaemrreports.data.converter.ARVsDataConverter.convert(ARVsDataConverter.java:27)
	at org.openmrs.module.reporting.data.DataUtil.convertData(DataUtil.java:57)
	at org.openmrs.module.reporting.dataset.definition.evaluator.PatientDataSetEvaluator.evaluate(PatientDataSetEvaluator.java:99)
	at org.openmrs.module.reporting.dataset.definition.evaluator.PatientDataSetEvaluator.evaluate(PatientDataSetEvaluator.java:42)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.executeEvaluator(BaseDefinitionService.java:369)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.evaluateBypassingExclusionOfTestPatients(BaseDefinitionService.java:272)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.evaluate(BaseDefinitionService.java:229)
	at org.openmrs.module.reporting.dataset.definition.service.DataSetDefinitionServiceImpl.evaluate(DataSetDefinitionServiceImpl.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.cache.interceptor.CacheInterceptor$1.invoke(CacheInterceptor.java:52)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:303)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:122)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy161.evaluate(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:201)
	at com.sun.proxy.$Proxy291.evaluate(Unknown Source)
	at org.openmrs.module.reporting.report.definition.evaluator.DefaultReportDefinitionEvaluator.evaluate(DefaultReportDefinitionEvaluator.java:78)
	at org.openmrs.module.reporting.report.definition.evaluator.DefaultReportDefinitionEvaluator.evaluate(DefaultReportDefinitionEvaluator.java:35)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.executeEvaluator(BaseDefinitionService.java:369)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.evaluateBypassingExclusionOfTestPatients(BaseDefinitionService.java:272)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.evaluate(BaseDefinitionService.java:229)
	at org.openmrs.module.reporting.report.definition.service.ReportDefinitionServiceImpl.evaluate(ReportDefinitionServiceImpl.java:168)
	at org.openmrs.module.reporting.report.definition.service.ReportDefinitionServiceImpl.evaluate(ReportDefinitionServiceImpl.java:37)
	at org.openmrs.module.reporting.definition.service.BaseDefinitionService.evaluate(BaseDefinitionService.java:353)
	at org.openmrs.module.reporting.report.definition.service.ReportDefinitionServiceImpl.evaluate(ReportDefinitionServiceImpl.java:160)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.cache.interceptor.CacheInterceptor$1.invoke(CacheInterceptor.java:52)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:303)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:122)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy150.evaluate(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:201)
	at com.sun.proxy.$Proxy290.evaluate(Unknown Source)
	at org.openmrs.module.reporting.report.service.ReportServiceImpl.runReport(ReportServiceImpl.java:471)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:201)
	at com.sun.proxy.$Proxy223.runReport(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.cache.interceptor.CacheInterceptor$1.invoke(CacheInterceptor.java:52)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:303)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.openmrs.aop.LoggingAdvice.invoke(LoggingAdvice.java:122)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy224.runReport(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:201)
	at com.sun.proxy.$Proxy224.runReport(Unknown Source)
	at org.openmrs.module.reporting.report.task.RunQueuedReportsTask.executeTask(RunQueuedReportsTask.java:59)
	at org.openmrs.module.reporting.report.task.ReportingTask.run`

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java (hunk 7 above)
Comment (solemabrothers) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupANCRegister2019.java]: Done
